libdnf/goal/Goal.cpp

```diff
@@ -793,6 +793,92 @@ Goal::favor(DnfPackage *pkg)
     queue_push2(&pImpl->staging, SOLVER_SOLVABLE|SOLVER_FAVOR, dnf_package_get_id(pkg));
 }
 
+void
+Goal::add_exclude_from_weak(const DnfPackageSet & pset)
+{
+    pImpl->exclude_from_weak += pset;
+}
+
+void
+Goal::add_exclude_from_weak(DnfPackage *pkg)
+{
+    // ensure that the map has a corrent size before set to prevent memory corruption
+    map_grow(pImpl->exclude_from_weak.getMap(), dnf_sack_get_pool(pImpl->sack)->nsolvables);
+    pImpl->exclude_from_weak.set(pkg);
+}
+
+void
+Goal::reset_exclude_from_weak()
+{
+    pImpl->exclude_from_weak.clear();
+}
+
+void
+Goal::exclude_from_weak_autodetect()
+{
+    Query installed_query(pImpl->sack, Query::ExcludeFlags::IGNORE_EXCLUDES);
+    installed_query.installed();
+    if (installed_query.empty()) {
+        return;
+    }
+    Query base_query(pImpl->sack);
+    base_query.apply();
+    auto * installed_pset = installed_query.getResultPset();
+    Id installed_id = -1;
+
+    std::vector<const char *> installed_names;
+    installed_names.reserve(installed_pset->size());
```

Comment:
Reserve the size + 1 for the nullptr that you push_back at the end?

Reply:
Thanks, it is correct suggestion.